build-logic/cleanup/src/main/java/gradlebuild/cleanup/KillLeakingJavaProcesses.java

```diff
@@ -0,0 +1,248 @@
+         */
+        String classpath;
+
+        private JavaProcessInfo(String pid, String javaCommand, String otherCommands) {
+            this.pid = pid;
+            commands.add(javaCommand);
+            Stream.of(otherCommands.split("\\s"))
+                .map(String::trim)
+                .filter(s -> !s.isEmpty())
+                .forEach(commands::add);
+            for (int i = 0; i < commands.size(); ++i) {
+                String command = commands.get(i);
+                if (MAIN_CLASS_PATTERN.matcher(command).matches()) {
+                    mainClass = command;
+                }
+                if (i != commands.size() - 1 && "-classpath".equals(command) || "-cp".equals(command)) {
+                    classpath = commands.get(i + 1);
+                    if (classpath.startsWith("\"")) {
+                        classpath = classpath.substring(1, classpath.length());
+                    }
+                    if (classpath.endsWith("\"")) {
+                        classpath = classpath.substring(0, classpath.length() - 1);
+                    }
+                }
+            }
+        }
+
+        static List<JavaProcessInfo> parsePsOutput(String psOutput, boolean isWindows) {
+            return psOutput.lines()
+                .map(line -> JavaProcessInfo.parseLine(line, isWindows))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+        }
+
+        /**
+         * Parse a jps output line and get the Java process info.
+         *
+         * @return null if the line doesn't match the pattern.
+         */
+        private static JavaProcessInfo parseLine(String line, boolean isWindows) {
+            Pattern pattern = isWindows ? WINDOWS_JAVA_PROCESS_PATTERN : UNIX_JAVA_PROCESS_PATTERN;
+            Matcher matcher = pattern.matcher(line.trim());
+            if (matcher.find()) {
+                if (isWindows) {
+                    return new JavaProcessInfo(matcher.group(4), matcher.group(1), matcher.group(3));
+                } else {
+                    return new JavaProcessInfo(matcher.group(1), matcher.group(2), matcher.group(3));
+                }
+            } else {
+                return null;
+            }
+        }
+
+        private boolean isLeakingJavaProcess() {
+            if (mainClass == null) {
+                return false;
+            }
+            if (mainClass.endsWith(".GradleWorkerMain") || mainClass.endsWith(".KotlinCompileDaemon")) {
+                return true;
+            } else if (mainClass.endsWith(".GradleDaemon")) {
+                return classpath != null && classpath.startsWith(WORKING_DIRECTORY_PATH);
+            } else {
+                return false;
+            }
+        }
+
+        private void kill() {
+            System.out.println("Killing process " + pid);
+            ExecResult execResult = isWindows() ? run("taskkill.exe", "/F", "/T", "/PID", pid) : run("kill", "-9", pid);
```

Comment:
Can we do a nice kill first, and only afterwards a force kill? So the daemons can shut down orderly?

Reply:
How? With `ProcessHandle.destroy()`?

Reply:
Use `kill <pid>` instead of `kill`, and then check if the processes we killed actually terminated within 5 seconds? And if they didn't, do the `kill -9 <pid>` ( or the `taskkill /F`).